dfusion_rust_core/src/models/solution.rs

```diff
@@ -17,12 +16,65 @@ pub struct Solution {
 impl Solution {
     pub fn trivial(num_orders: usize) -> Self {
         Solution {
-            objective_value: Some(U256::zero()),
             prices: vec![0; TOKENS as usize],
             executed_buy_amounts: vec![0; num_orders],
             executed_sell_amounts: vec![0; num_orders],
         }
     }
+
+    /// Returns the token price for token or `None` if the token is invalid.
+    fn price(&self, token: u16) -> Option<u128> {
+        self.prices.get(token as usize).cloned()
+    }
+
+    /// Compute the result of the objective function for the solution. This is
+    /// currently defined as *U - dUT* where *U* is the total utility of the
+    /// trade and *dUT* is the total disreguarded utility of touched orders.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` for any solution that is positive valid solution.
+    /// Returns `None` for invalid solutions or orders. A solution might be
+    /// invalid if:
+    /// - The number of orders do not match the solution buy/sell amounts
+    /// - The solution values lead to invalid utility or disregarded utility
+    ///   results
+    /// - The orders reference tokens that are not priced by the solution
+    pub fn objective_value(&self, orders: &[Order]) -> Option<U256> {
+        if orders.len() != self.executed_buy_amounts.len()
+            || orders.len() != self.executed_sell_amounts.len()
+        {
+            return None;
+        }
+
+        // note that we don't actually have to do the same checking here (such as
+        // account balances and such) as we do in the smart contract - we only
+        // care if the solution is valid given the provided orders
+
+        let mut u = U256::zero();
+        let mut du = U256::zero();
+
+        for (i, order) in orders.iter().enumerate() {
+            let buy_price = self.price(order.buy_token)?;
+            let sell_price = self.price(order.sell_token)?;
+            let exec_buy_amount = self.executed_buy_amounts[i];
+            let exec_sell_amount = self.executed_sell_amounts[i];
+
+            if exec_buy_amount == 0 && exec_sell_amount == 0 {
+                // this order was not touched, so skip
+                continue;
+            }
+
+            u = u.checked_add(order.utility(buy_price, sell_price, exec_buy_amount)?)?;
+            du = du.checked_add(order.disregarded_utility(
+                buy_price,
+                sell_price,
+                exec_sell_amount,
+            )?)?;
```

Comment:
These actually turn out to look super clean when all the ugly calculations are buried deep inside the `utility` and `disregarded utility` functions!